### PR TITLE
Warn about loss of precision in FDBSCAN-DenseBox

### PR DESCRIPTION
**Pull request by aprokop**
For some problems and eps, FDBSCAN-DenseBox may suffer from severe loss
of precision when computing cellBox(). This may result in wrong DBSCAN
results.

Initially, the problem was observed on a full NGSIM dataset:
```
$ ./ArborX_DBSCAN.exe --binary \
    --filename /data/data_and_logs/2021_02_20_dbscan_datasets/mustafa_2019/ngsim.arborx \
    --print-dbscan-timers --core-min-size 2 --eps 1.0 --max-num-points 90000 \
    --impl fdbscan-densebox --verify
<...>
Core point is marked as noise: 41888 [-1]
Core point is marked as noise: 17027 [-1]
Core point is marked as noise: 43104 [-1]
```
Tracking down the issue revealed that a point 2279 (part of a dense
cell) was finding point 10369 (not part of a dense cell), but the
reverse was not true (i.e., 10369 was not finding the dense cell that
2279 was in). Studying the dense boxes, I observed that the problematic
box had bounds
```
(6452399.000000,1872182.375000,0.000000)
(6452399.000000,1872183.000000,0.577350)
```
Note the same x coordinate for both min and max corner. The h in this
run was ~0.57. So it completely got lost. Even Y coordinate lost some
precision, though not as much.

I'm not sure how robust this is. I also don't know what the best way to indicate this to user. My issue with using `ARBORX_ASSERT` is that when a user compiles ArborX in release mode, this will never get noticed. On the other hand, hardcoding a `printf` statement contradicts xSDK policy on having no hardcoded printouts.

### CODE REVIEW

**Comment by aprokop** [comment]
Currently I don't have a better idea, so lets go with the current version.

**Comment by aprokop** [comment]
Please review.

**Comment by dalg24** [comment]
> My issue with using ARBORX_ASSERT is that when a user compiles ArborX in release mode, this will never get noticed.

As currently implemented you would always get the runtime error
https://github.com/arborx/ArborX/blob/6aca3b81a6537c33efd91d37429bb10ed8421fd0/src/details/ArborX_Exception.hpp#L33-L38

File: src/details/ArborX_DetailsFDBSCANDenseBox.hpp [hunk 1]
```diff
@@ -51,6 +51,26 @@ struct CartesianGrid
...
+    float constexpr eps = 5 * std::numeric_limits<float>::epsilon();
+    if (std::abs(_h / min_corner[0]) < eps ||
+        std::abs(_h / min_corner[1]) < eps ||
+        std::abs(_h / min_corner[2]) < eps ||
+        std::abs(_h / max_corner[0]) < eps ||
+        std::abs(_h / max_corner[1]) < eps ||
+        std::abs(_h / max_corner[2]) < eps)
+      printf("ArborX warning: FDBSCAN-DenseBox algorithm may experience loss "
+             "of precision, producing wrong results. Please consider switching "
+             "to FDBSCAN.\n");
```
Comment (dalg24) [inline src/details/ArborX_DetailsFDBSCANDenseBox.hpp:73]: Any good reason to prefer `printf` to the C++ I/O facility.  Also consider printing to the standard error stream.

**Comment by aprokop** [comment]
> As currently implemented you would always get the runtime error

Correct. I don't see a graceful recovery from what could be erroneous results.

File: src/details/ArborX_DetailsFDBSCANDenseBox.hpp (hunk 1 above)
Comment (aprokop) [reply src/details/ArborX_DetailsFDBSCANDenseBox.hpp:73]: No good reason. Would like to change?

Comment (aprokop) [reply src/details/ArborX_DetailsFDBSCANDenseBox.hpp]: I switched to `std::cerr`.

**Comment by aprokop** [comment]
@dalg24 Can we merge this PR in its current form? It's not robust, but improves *status quo*.

**Comment by aprokop** [comment]
Please review.

File: src/details/ArborX_DetailsCartesianGrid.hpp [hunk 2]
```diff
@@ -118,6 +121,23 @@ struct CartesianGrid
...
     }
+
+    // Catch a potential loss of precision that may happen in cellBox() and can
+    // lead to wrong results.
+    //
+    // The machine precision by itself is not sufficient. In some experiments
+    // run with a full NGSIM datasets, values below 3 could still produce wrong
+    // results. This may still not be conservative enough, but all runs passed
+    // verification when this warning was not triggered.
+    float constexpr eps = 5 * std::numeric_limits<float>::epsilon();
```
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:132]:
```suggestion
    constexpr auto eps = 5 * std::numeric_limits<float>::epsilon();
```
We use "constexpr West" and here `float` is redundant

File: src/details/ArborX_DetailsCartesianGrid.hpp [hunk 3]
```diff
@@ -118,6 +121,23 @@ struct CartesianGrid
...
+    for (int d = 0; d < DIM; ++d)
+    {
+      if (std::abs(_h[d] / min_corner[d]) < eps)
+        std::cerr << "ArborX warning: FDBSCAN-DenseBox algorithm may "
+                     "experience loss of precision, producing wrong results. "
+                     "Please consider switching to FDBSCAN."
+                  << std::endl;
```
Comment (dalg24) [inline src/details/ArborX_DetailsCartesianGrid.hpp:139]: Fine but the will be the first occurrence of printing to `cerr` or `cout` and I am not looking forward to introducing some mechanism to not print the warnings down the road.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Must fix the style

**Review by dalg24 — APPROVED** [review]

**Pull request merged**